libs/remote_storage/src/lib.rs

```diff
@@ -313,6 +337,23 @@ impl<Other: RemoteStorage> GenericRemoteStorage<Arc<Other>> {
         }
     }
 
+    // See [`RemoteStorage::list_streaming`].
+    pub async fn list_streaming<'a>(
+        &'a self,
+        prefix: Option<&'a RemotePath>,
+        mode: ListingMode,
+        max_keys: Option<NonZeroU32>,
+        cancel: &'a CancellationToken,
+    ) -> impl Stream<Item = anyhow::Result<Listing, DownloadError>> + 'a {
+        match self {
+            Self::LocalFs(s) => Box::pin(s.list_streaming(prefix, mode, max_keys, cancel).await)
+                as Pin<Box<dyn Stream<Item = anyhow::Result<Listing, DownloadError>>>>,
+            Self::AwsS3(s) => Box::pin(s.list_streaming(prefix, mode, max_keys, cancel).await),
+            Self::AzureBlob(s) => Box::pin(s.list_streaming(prefix, mode, max_keys, cancel).await),
+            Self::Unreliable(s) => Box::pin(s.list_streaming(prefix, mode, max_keys, cancel).await),
+        }
```

Comment:
Did you check if the compilation time balloons up if you use two levels of either?

Reply:
What do you mean by this, could you expand? The Box is needed because the implemented types are different for each branch. The box is a way to make the types match.

Reply:
With four eithers I meant:

```
use futures::Either;

match self {
    LocalFs(s) => Either::Left(Either::Left(s.list_streaming(...))),
    AwsS3(s) => Either::Left(Either::Right(s.list_streaming(...))),
    AzureBlob(s) => Either::Right(Either::Left(s.list_streaming(...))),
    Unreliable(s) => Either::Right(Either::Right(s.list_streaming(...))),
}
```

This would be another way of making the types match. But yeah, might be waste of time.